runtime/src/block_min_prioritization_fee_cache.rs

```diff
@@ -0,0 +1,595 @@
+use {
+    crate::{
+        block_min_prioritization_fee::*,
+        block_min_prioritization_fee_cache_query::BlockMinPrioritizationFeeCacheQuery,
+        block_min_prioritization_fee_cache_update::BlockMinPrioritizationFeeCacheUpdate,
+    },
+    solana_sdk::{
+        clock::Slot, pubkey::Pubkey, saturating_add_assign, transaction::SanitizedTransaction,
+    },
+    std::collections::HashMap,
+};
+
+/// The maximum number of blocks to keep in `BlockMinPrioritizationFeeCache`; States from
+/// up to 150 recent blocks should be sufficient to estimate minimal prioritization fee to
+/// land transactions to current block.
+const NUMBER_OF_RECENT_BLOCKS: usize = 150;
+
+/// Holds up to NUMBER_OF_RECENT_BLOCKS recent block's min prioritization fee for block,
+/// and for each writable accounts per block.
+pub struct BlockMinPrioritizationFeeCache {
+    cache: HashMap<Slot, BlockMinPrioritizationFee>,
+    // metrics counts
+    successful_transaction_update_count: u64,
+    fail_get_transaction_priority_details_count: u64,
+    fail_get_transaction_account_locks_count: u64,
+    fail_finalize_block_not_found: u64,
+}
+
+impl Default for BlockMinPrioritizationFeeCache {
+    fn default() -> Self {
+        Self::new(NUMBER_OF_RECENT_BLOCKS)
+    }
+}
+
+impl BlockMinPrioritizationFeeCache {
+    pub fn new(capacity: usize) -> Self {
+        BlockMinPrioritizationFeeCache {
+            cache: HashMap::with_capacity(capacity),
+            //            ..default()
+            successful_transaction_update_count: 0,
+            fail_get_transaction_priority_details_count: 0,
+            fail_get_transaction_account_locks_count: 0,
+            fail_finalize_block_not_found: 0,
+        }
+    }
+
+    fn get_block_min_prioritization_fee(&self, slot: &Slot) -> Option<&BlockMinPrioritizationFee> {
+        self.cache.get(slot)
+    }
+
+    fn get_mut_block_min_prioritization_fee(
+        &mut self,
+        slot: &Slot,
+    ) -> Option<&mut BlockMinPrioritizationFee> {
+        self.cache.get_mut(slot)
+    }
+
+    fn get_or_add_mut_block_min_prioritization_fee(
+        &mut self,
+        slot: &Slot,
+    ) -> &mut BlockMinPrioritizationFee {
+        self.cache
+            .entry(*slot)
+            .or_insert_with(BlockMinPrioritizationFee::default)
+    }
+
+    fn report_metrics_for_slot(&mut self, slot: Slot) {
+        // report current min fees for slot
+        if let Some(block) = self.get_block_min_prioritization_fee(&slot) {
+            datapoint_info!(
+                "block_min_prioritization_fee",
+                ("slot", slot as i64, i64),
+                ("entity", "block", String),
+                (
+                    "min_prioritization_fee",
+                    block.get_block_fee().unwrap_or(0) as i64,
+                    i64
+                ),
+            );
+            for (account_key, fee) in block.get_account_fees() {
```

Comment:
Seems like this could be really noisy for metrics... Or do you expect very few writable accounts to end up in this map?

Reply:
Yep, since accounts have fee less or equal to block_fee will be removed from cache during finalize call, I expect only few accounts in each block to report. But I dont have good number to prove it yet. May be just report top 10 accounts? Then I have to sort first.